hazelcast/src/main/java/com/hazelcast/internal/usercodedeployment/impl/ClassDataProvider.java

```diff
@@ -51,38 +53,73 @@ public ClassDataProvider(UserCodeDeploymentConfig.ProviderMode providerMode,
     }
 
     public ClassData getClassDataOrNull(String className) {
-        byte[] bytecode = loadBytecodeFromClientCache(className);
-        if (bytecode != null) {
-            return new ClassData(bytecode);
+        ClassData classData = loadBytecodesFromClientCache(className);
+        if (classData != null) {
+            return classData;
         }
         if (providerMode == UserCodeDeploymentConfig.ProviderMode.OFF) {
             return null;
         }
-        ClassData classData = null;
-        bytecode = loadBytecodeFromParent(className);
-        if (bytecode == null && providerMode == UserCodeDeploymentConfig.ProviderMode.LOCAL_AND_CACHED_CLASSES) {
-            bytecode = loadBytecodeFromCache(className);
-        }
-        if (bytecode != null) {
-            classData = new ClassData(bytecode);
+        classData = loadBytecodesFromParent(className);
+        if (classData == null && providerMode == UserCodeDeploymentConfig.ProviderMode.LOCAL_AND_CACHED_CLASSES) {
+            classData = loadBytecodesFromCache(className);
         }
         return classData;
     }
 
-    private byte[] loadBytecodeFromCache(String className) {
-        ClassSource classSource = classSourceMap.get(className);
+    private ClassData loadBytecodesFromCache(String className) {
+        ClassSource classSource = classSourceMap.get(ClassLocator.extractOuterClassName(className));
         if (classSource == null) {
             return null;
         }
-        return classSource.getBytecode();
+        return classSource.getClassData(className);
     }
 
-    private byte[] loadBytecodeFromClientCache(String className) {
-        ClassSource classSource = clientClassSourceMap.get(className);
+    private ClassData loadBytecodesFromClientCache(String className) {
+        ClassSource classSource = clientClassSourceMap.get(ClassLocator.extractOuterClassName(className));
         if (classSource == null) {
             return null;
         }
-        return classSource.getBytecode();
+        return classSource.getClassData(className);
+    }
+
+    private ClassData loadBytecodesFromParent(String className) {
+        byte[] requestedClassDefinition = loadBytecodeFromParent(className);
+        if (requestedClassDefinition == null) {
+            return null;
+        }
+        HashMap<String, byte[]> classDefinitions = new HashMap<String, byte[]>();
+//        classDefinitions.put(className, requestedClassDefinition);
```

Comment:
This line should have been deleted. This was a failed attempt. I was trying to keep all class definitions in one map. But later realised, it will cause problems. But this resembles the my intent after 3.9. I think you meant that when saying ` you don't need this at all.`